fv-tests/server_test.go

```diff
@@ -0,0 +1,710 @@
+		}
+		if cache != nil {
+			cacheCancel()
+		}
+	})
+
+	It("should not disconnect a responsive client", func() {
+		// Start a real client, which will respond correctly to pings.
+		clientCxt, clientCancel := context.WithCancel(context.Background())
+		recorder := &stateRecorder{
+			kvs: map[string]api.Update{},
+		}
+		client := syncclient.New(
+			serverAddr,
+			"test-version",
+			"test-host",
+			"test-info",
+			recorder,
+		)
+		err := client.StartContext(clientCxt)
+		Expect(err).NotTo(HaveOccurred())
+		defer func() {
+			clientCancel()
+			client.Finished.Wait()
+		}()
+
+		// Wait until we should have been dropped if we were unresponsive.
+		time.Sleep(1 * time.Second)
+
+		// Then send an update.
+		cache.OnStatusUpdated(api.InSync)
+		Eventually(recorder.Status).Should(Equal(api.InSync))
+	})
+
+	Describe("with a raw connection", func() {
+		var rawConn net.Conn
+		var w *gob.Encoder
+		var r *gob.Decoder
+
+		BeforeEach(func() {
+			var err error
+			rawConn, err = net.DialTimeout("tcp", serverAddr, 10*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+
+			w = gob.NewEncoder(rawConn)
+			r = gob.NewDecoder(rawConn)
+		})
+
+		AfterEach(func() {
+			err := rawConn.Close()
+			if err != nil {
+				log.WithError(err).Info("Error recorded while closing conn.")
+			}
+		})
+
+		expectDisconnection := func(after time.Duration) {
+			var envelope syncproto.Envelope
+			startTime := time.Now()
+			for {
+				err := r.Decode(&envelope)
+				if err != nil {
+					return // Success!
+				}
+				if time.Since(startTime) > after {
+					Fail("Client should have been disconnected by now")
+				}
+			}
+			expectGaugeValue("typha_connections_active", 0.0)
+		}
+
+		It("should clean up if the hello doesn't get sent", func() {
+			expectGaugeValue("typha_connections_active", 1.0)
+			rawConn.Close()
+			expectGaugeValue("typha_connections_active", 0.0)
+		})
+
+		Describe("After sending Hello", func() {
+			BeforeEach(func() {
+				err := w.Encode(syncproto.Envelope{
+					Message: syncproto.MsgClientHello{
+						Hostname: "me",
+						Version:  "test",
+						Info:     "test info",
+					},
+				})
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			It("should disconnect a client that sends a nil update", func() {
+				var envelope syncproto.Envelope
+				err := w.Encode(envelope)
+				Expect(err).NotTo(HaveOccurred())
+				expectDisconnection(100 * time.Millisecond)
+			})
+
+			It("should disconnect a client that sends an unexpected update", func() {
+				var envelope syncproto.Envelope
+				envelope.Message = 42
+				err := w.Encode(envelope)
+				Expect(err).NotTo(HaveOccurred())
+				expectDisconnection(100 * time.Millisecond)
+			})
+
+			It("should disconnect a client that sends a garbage update", func() {
+				rawConn.Write([]byte("dsjfkldjsklfajdskjfk;dajskfjaoirefmuweioufijsdkfjkdsjkfjasd;"))
+				// We get dropped by the pong timeout since the gob decoder is confused.
+				expectDisconnection(time.Second)
```

Comment:
I'll beef up the comment on the line above, which I thought already explained this.